File: carPooling.cpp
class Solution {
public:
    bool carPooling(vector<vector<int>>& trips, int capacity) {
        
        int n = trips.size(), tripEnd = 0;

        for(int i = 0; i < n; i++)
            tripEnd = max(trips[i][2], tripEnd);

        tripEnd++;

        vector<int> totalPassengers(tripEnd, 0);

        for(int i = 0; i < n; i++)
        {
            totalPassengers[trips[i][1]] += trips[i][0];

            totalPassengers[trips[i][2]] -= trips[i][0];
        }

        for(int i = 1; i < tripEnd; i++)
        {
            totalPassengers[i] += totalPassengers[i - 1];

            if(totalPassengers[i] > capacity) return false;
        }

        if(totalPassengers[0] > capacity) return false;

        return true;
    }
};

File: progress_sheet/carPooling.py
class Solution:
    def carPooling(self, trips: List[List[int]], capacity: int) -> bool:

        # Line sweeping technique

        max_range = 0
        prefix_trips = []

        for trip in trips:
            max_trip = max(trip[1], trip[2])
            max_range = max(max_range, max_trip)
        
        prefix_trips = [0] * (max_range + 2)
        
        # Mark pick up and drop off points
        for trip in trips:
            prefix_trips[trip[1]] += trip[0]
            prefix_trips[trip[2]] -= trip[0]
        
        # sweep the marked points by taking the cumulative sum
        # and check if we have enough seats
        for i in range(len(prefix_trips)):
            prefix_trips[i] += prefix_trips[i - 1]

            if prefix_trips[i] > capacity:
                return False
        
        return True
